File: src/main/java/minefarts/smarttube/utils/next/v1/WatchNextApi.java
package minefarts.smarttube.utils.next.v1;

import minefarts.smarttube.google.common.converters.jsonpath.WithJsonPath;
import minefarts.smarttube.utils.next.v1.result.WatchNextResultContinuation;
import minefarts.smarttube.utils.next.v1.result.WatchNextResult;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.Header;
import retrofit2.http.Headers;
import retrofit2.http.POST;

/**
 * For signed users!
 */
@WithJsonPath
public interface WatchNextApi {

    @Headers("Content-Type: application/json")
    @POST("https://www.youtube.com/youtubei/v1/next")
    Call<WatchNextResult> getWatchNextResult(@Body String watchNextQuery);

    @Headers("Content-Type: application/json")
    @POST("https://www.youtube.com/youtubei/v1/next")
    Call<WatchNextResultContinuation> continueWatchNextResult(@Body String watchNextQuery);

    @Headers("Content-Type: application/json")
    @POST("https://www.youtube.com/youtubei/v1/next")
    Call<WatchNextResultContinuation> continueWatchNextResult(
        @Body String suggestQuery, 
        @Header("X-Goog-Visitor-Id") String visitorId
    );
    
}

File: src/main/java/minefarts/smarttube/utils/next/v2/WatchNextApi.kt
package minefarts.smarttube.utils.next.v2

import minefarts.smarttube.google.common.converters.gson.WithGson
import minefarts.smarttube.utils.next.v2.gen.UnlocalizedTitleResult
import minefarts.smarttube.utils.next.v2.gen.WatchNextResult
import minefarts.smarttube.utils.next.v2.gen.WatchNextResultContinuation

import retrofit2.Call
import retrofit2.http.*

import okhttp3.RequestBody

@WithGson
public interface WatchNextApi {

    @Headers("Content-Type: application/json")
    @POST("https://www.youtube.com/youtubei/v1/next")
    fun getWatchNextResult(@Body watchNextQuery: String): Call<WatchNextResult?>

    @Headers("Content-Type: application/json")
    @POST("https://www.youtube.com/youtubei/v1/next")
    fun getWatchNextResult(
        @Body watchNextQuery: RequestBody,
        @Header("X-Goog-Visitor-Id") visitorId: String
    ): Call<WatchNextResult?>

    @Headers("Content-Type: application/json")
    @POST("https://www.youtube.com/youtubei/v1/next")
    fun continueWatchNextResult(@Body watchNextQuery: String): Call<WatchNextResultContinuation?>

    @Headers("Content-Type: application/json")
    @POST("https://www.youtube.com/youtubei/v1/next")
    fun continueWatchNextResult(
        @Body watchNextQuery: String, 
        @Header("X-Goog-Visitor-Id") visitorId: String
    ): Call<WatchNextResultContinuation?>

    @GET("https://www.youtube.com/oembed")
    fun getUnlocalizedTitle(@Query("url") url: String): Call<UnlocalizedTitleResult?>

}